### PR TITLE
Generator: items in the VariableInfoWithType array may not be in the correct order

### PR DESCRIPTION
**Pull request by agarny**
Addresses issue #597.

### CODE REVIEW

**Comment by nickerso** [comment]
Just to reiterate earlier requests to non contaminate PRs with unrelated changes :) the clang tidy changes especially make it difficult to see how this will fix #597...

**Comment by agarny** [comment]
I know, but... I can't compile without PR #596, hence this PR will have to be merged in after #596. And, once it has been done, you will be able to see what this PR changes... :)

**Review by nickerso — DISMISSED** [review]
This also fixes the variable ordering issue in #598.

**Comment by agarny** [comment]
Ok, this PR now only contains changes that address issue #597. @nickerso, @hsorby and @kerimoyle, feel free to review and merge in.

File: src/generator.cpp [hunk 1]
```diff
@@ -1373,14 +1377,31 @@ bool Generator::GeneratorImpl::compareVariablesByName(const GeneratorInternalVar
...
+    return variable->mType == GeneratorInternalVariable::Type::STATE;
+}
+
+bool Generator::GeneratorImpl::isConstantOrAlgebraicVariable(const GeneratorInternalVariablePtr &variable)
+{
+    return (variable->mType == GeneratorInternalVariable::Type::CONSTANT)
+           || (variable->mType == GeneratorInternalVariable::Type::COMPUTED_TRUE_CONSTANT)
+           || (variable->mType == GeneratorInternalVariable::Type::COMPUTED_VARIABLE_BASED_CONSTANT)
+           || (variable->mType == GeneratorInternalVariable::Type::ALGEBRAIC);
+}
```
Comment (kerimoyle) [inline src/generator.cpp:1392]: What's the difference between these types?  `CONSTANT` and `COMPUTED_TRUE_CONSTANT`? And `COMPUTED_VARIABLE_BASED_CONSTANT` and later you have `COMPUTED_CONSTANT` as well?

Comment (agarny) [reply src/generator.cpp:1392]: They are internal types to the generator to determine our different types of constants:
 - `x = 3` --> `CONSTANT`;
 - `y = 3 + 5` --> `COMPUTED_TRUE_CONSTANT`; and
 - `z = a + b` with both `a` and `b` constants --> `COMPUTED_VARIABLE_BASED_CONSTANT`.

However, from the outside (e.g. OpenCOR), we only want to know whether a constant is a "true" constant or a computed constant, hence:
 - internal `CONSTANT` --> external `CONSTANT`;
 - internal `COMPUTED_TRUE_CONSTANT` --> external `COMPUTED_CONSTANT`; and
 - internal `COMPUTED_VARIABLE_BASED_CONSTANT` --> external `COMPUTED_CONSTANT`.

HTH... :)

**Comment by agarny** [comment]
Ok, ready to go...

**Comment by kerimoyle** [comment]
This branch isn't up to date? or are there other changes in here too?

**Comment by agarny** [comment]
It is up to date. What makes you think that it may not be?

**Comment by kerimoyle** [comment]
Just seeing changes that I thought were in the other PR appearing here...

**Comment by agarny** [comment]
Nope, no chance.

**Review by kerimoyle — APPROVED** [review]

**Comment by agarny** [comment]
@nickerso, are you still happy with this PR? (The only difference between your last approval and now is that I had to merge the contents of `develop` into this PR.)

**Comment by nickerso** [comment]
Can approve via email? Yes, I still approve.

**Comment by agarny** [comment]
> Can approve via email? Yes, I still approve.

GitHub says nooooo.

**Review by nickerso — APPROVED** [review]

**Pull request merged**